File: item_decor/src/main/java/me/simple/itemdecor/LinearItemDecor.java
package me.simple.itemdecor;

import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.RectF;
import android.view.View;

import androidx.recyclerview.widget.RecyclerView;

public class LinearItemDecor implements IFilter<LinearItemDecor> {

    private Paint mPaint;
    private RectF mLineRect = new RectF();

    public static final int HORIZONTAL = 0;
    public static final int VERTICAL = 1;
    private int mOrientation = VERTICAL;

    /**
     * ItemDecoration的颜色
     */
    private int mColor = Color.GRAY;
    /**
     *
     */
    private int mHeight = 1;
    private int mWidth = 1;
    /**
     *
     */
    private float mMarginLeft;
    private float mMarginTop;
    private float mMarginRight;
    private float mMarginBottom;

    /**
     *
     */
    private FilterFunc mFilterFunc;
    /**
     * 是否保存最后一个ItemDecoration
     */
    private boolean mRetainLast = false;

    public LinearItemDecor() {
        this(new Paint(Paint.ANTI_ALIAS_FLAG));
    }

    public LinearItemDecor(Paint paint) {
        paint.setColor(mColor);
        this.mPaint = paint;
    }

    public LinearItemDecor(int mOrientation) {
        this(mOrientation, new Paint(Paint.ANTI_ALIAS_FLAG));
    }

    public LinearItemDecor(int mOrientation, Paint paint) {
        this.mOrientation = mOrientation;
        paint.setColor(mColor);
        this.mPaint = paint;
    }

    public LinearItemDecor setOrientation(int orientation) {
        this.mOrientation = orientation;
        return this;
    }

    public LinearItemDecor setColor(int color) {
        this.mColor = color;
        mPaint.setColor(mColor);
        return this;
    }

    public LinearItemDecor setHeight(int height) {
        this.mHeight = height;
        return this;
    }


    public LinearItemDecor setWidth(int width) {
        this.mWidth = width;
        return this;
    }

//    public LinearItemDecor setMargin(float margin) {
//        this.mMarginLeft = margin;
//        this.mMarginTop = margin;
//        this.mMarginRight = margin;
//        this.mMarginBottom = margin;
//        return this;
//    }

    public LinearItemDecor setMarginLeft(float marginLeft) {
        this.mMarginLeft = marginLeft;
        return this;
    }

    public LinearItemDecor setMarginTop(float marginTop) {
        this.mMarginTop = marginTop;
        return this;
    }

    public LinearItemDecor setMarginRight(float marginRight) {
        this.mMarginRight = marginRight;
        return this;
    }

    public LinearItemDecor setMarginBottom(float marginBottom) {
        this.mMarginBottom = marginBottom;
        return this;
    }

    public LinearItemDecor setMarginHorizontal(float margin) {
        this.mMarginLeft = margin;
        this.mMarginRight = margin;
        return this;
    }

    public LinearItemDecor setMarginVertical(float margin) {
        this.mMarginTop = margin;
        this.mMarginBottom = margin;
        return this;
    }

    @Override
    public LinearItemDecor filter(FilterFunc func) {
        this.mFilterFunc = func;
        return this;
    }

    /**
     * 保留最后一个ItemDecoration，默认不保留
     */
    public LinearItemDecor retainLast() {
        this.mRetainLast = true;
        return this;
    }

    public AbsItemDecor build() {
        return new AbsItemDecor() {
            @Override
            public void onDraw(Canvas canvas, int position, Rect bounds, View itemView,
                               RecyclerView parent, RecyclerView.State state) {
                if (mFilterFunc != null && mFilterFunc.exclude(position)) {
                    return;
                }

                if (!mRetainLast && parent.getAdapter() != null && position == parent.getAdapter().getItemCount() - 1) {
                    return;
                }

                if (mOrientation == VERTICAL) {
                    drawVertical(canvas, itemView, bounds, parent);
                } else {
                    drawHorizontal(canvas, itemView, bounds, parent);
                }
            }

            @Override
            public void onDrawOver(Canvas canvas, int position, Rect bounds, View itemView,
                                   RecyclerView parent, RecyclerView.State state) {

            }

            @Override
            public void setOutRect(Rect outRect, int position, View itemView,
                                   RecyclerView parent, RecyclerView.State state) {
                if (mFilterFunc != null && mFilterFunc.exclude(position)) {
                    outRect.set(0, 0, 0, 0);
                    return;
                }

                if (!mRetainLast && parent.getAdapter() != null && position == parent.getAdapter().getItemCount() - 1) {
                    outRect.set(0, 0, 0, 0);
                    return;
                }

                if (mOrientation == VERTICAL) {
                    outRect.set(0, 0, 0, mHeight);
                } else {
                    outRect.set(0, 0, mWidth, 0);
                }
            }
        };
    }

    private void drawVertical(Canvas canvas, View itemView, Rect bounds, RecyclerView parent) {
        canvas.save();
        int left;
        int right;
        if (parent.getClipToPadding()) {
            left = parent.getPaddingLeft();
            right = parent.getWidth() - parent.getPaddingRight();
            canvas.clipRect(left, parent.getPaddingTop(), right, parent.getHeight() - parent.getPaddingBottom());
        } else {
            left = 0;
            right = parent.getWidth();
        }
        left += mMarginLeft;
        right -= mMarginRight;
        int bottom = bounds.bottom + Math.round(itemView.getTranslationY());
        int top = bottom - mHeight;
        mLineRect.set(left, top, right, bottom);
        canvas.drawRect(mLineRect, mPaint);
        canvas.restore();
    }

    private void drawHorizontal(Canvas canvas, View itemView, Rect bounds, RecyclerView parent) {
        canvas.save();
        int top;
        int bottom;
        if (parent.getClipToPadding()) {
            top = parent.getPaddingTop();
            bottom = parent.getHeight() - parent.getPaddingBottom();
            canvas.clipRect(parent.getPaddingLeft(), top, parent.getWidth() - parent.getPaddingRight(), bottom);
        } else {
            top = 0;
            bottom = parent.getHeight();
        }
        top += mMarginTop;
        bottom -= mMarginBottom;
        int right = bounds.right + Math.round(itemView.getTranslationX());
        int left = right - mWidth;
        mLineRect.set(left, top, right, bottom);
        canvas.drawRect(mLineRect, mPaint);
        canvas.restore();
    }

}

File: item_decor/src/main/java/me/simple/itemdecor/LinearItemDecor.kt
package me.simple.itemdecor

import android.graphics.*
import android.view.View
import androidx.recyclerview.widget.RecyclerView
import java.util.*
import kotlin.math.roundToInt

class LinearItemDecor : AbsItemDecor(), IFilter<LinearItemDecor> {

    companion object {
        const val HORIZONTAL = 0
        const val VERTICAL = 1

        fun new(
            color: Int = Color.LTGRAY,
            size: Int = 1,
            margin: Float = 0f,
            retainLast: Boolean = false,
            orientation: Int = VERTICAL
        ) = LinearItemDecor().apply {
            this.color = color
            this.size = size
            this.margin = margin
            this.retainLast = retainLast
            this.orientation = orientation
        }

        fun new(
            color: Int = Color.LTGRAY,
            size: Int = 1,
            marginStart: Float = 0f,
            marginEnd: Float = 0f,
            retainLast: Boolean = false,
            orientation: Int = VERTICAL
        ) = LinearItemDecor().apply {
            this.color = color
            this.size = size
            this.marginStart = marginStart
            this.marginEnd = marginEnd
            this.retainLast = retainLast
            this.orientation = orientation
        }
    }

    private val mLineRect = RectF()

    //默认的paint，可以自定义替换掉
    var paint: Paint = Paint(Paint.ANTI_ALIAS_FLAG)

    //LinearLayoutManager的方向
    var orientation = VERTICAL

    //divider颜色
    var color = Color.LTGRAY
        set(value) {
            paint.color = value
            field = value
        }

    //divider的大小
    var size: Int = 1

    //
    var marginStart = 0f
    var marginEnd = 0f

    var margin = 0f
        set(value) {
            this.marginStart = value
            this.marginEnd = value
            field = value
        }


    //是否保存最后一个divider
    var retainLast = false

    //要过滤掉的divider
    private var filterBlock: ((position: Int) -> Boolean)? = null

    override fun filter(block: (position: Int) -> Boolean): LinearItemDecor {
        this.filterBlock = block
        return this
    }

    override fun filter(vararg filters: Int) {
        filter { position ->
            filters.contains(position)
        }
    }

    override fun onDraw(
        canvas: Canvas,
        position: Int,
        bounds: Rect,
        itemView: View,
        parent: RecyclerView,
        state: RecyclerView.State
    ) {
        //
        if (filterBlock?.invoke(position) == true) {
            return
        }
        //
        if (!retainLast && position == parent.adapter!!.itemCount - 1) {
            return
        }
        //
        if (orientation == VERTICAL) {
            drawVertical(canvas, itemView, bounds, parent)
        } else {
            drawHorizontal(canvas, itemView, bounds, parent)
        }
    }

    override fun setOutRect(
        outRect: Rect,
        position: Int,
        itemView: View,
        parent: RecyclerView,
        state: RecyclerView.State
    ) {
        //
        if (filterBlock?.invoke(position) == true) {
            outRect.set(0, 0, 0, 0)
            return
        }
        //
        if (!retainLast && position == parent.adapter!!.itemCount - 1) {
            outRect.set(0, 0, 0, 0)
            return
        }
        //
        if (orientation == VERTICAL) {
            outRect.set(0, 0, 0, size)
            outRect[0, 0, 0] = size
        } else {
            outRect.set(0, 0, size, 0)
        }
    }

    private fun drawVertical(canvas: Canvas, itemView: View, bounds: Rect, parent: RecyclerView) {
        canvas.save()
        var left: Int
        var right: Int
        if (parent.clipToPadding) {
            left = parent.paddingLeft
            right = parent.width - parent.paddingRight
            canvas.clipRect(left, parent.paddingTop, right, parent.height - parent.paddingBottom)
        } else {
            left = 0
            right = parent.width
        }
        left += marginStart.toInt()
        right -= marginEnd.toInt()
        val bottom = bounds.bottom + itemView.translationY.roundToInt()
        val top = bottom - size
        mLineRect[left.toFloat(), top.toFloat(), right.toFloat()] = bottom.toFloat()
        canvas.drawRect(mLineRect, paint)
        canvas.restore()
    }

    private fun drawHorizontal(canvas: Canvas, itemView: View, bounds: Rect, parent: RecyclerView) {
        canvas.save()
        var top: Int
        var bottom: Int
        if (parent.clipToPadding) {
            top = parent.paddingTop
            bottom = parent.height - parent.paddingBottom
            canvas.clipRect(parent.paddingLeft, top, parent.width - parent.paddingRight, bottom)
        } else {
            top = 0
            bottom = parent.height
        }
        top += marginStart.toInt()
        bottom -= marginEnd.toInt()
        val right = bounds.right + itemView.translationX.roundToInt()
        val left = right - size
        mLineRect[left.toFloat(), top.toFloat(), right.toFloat()] = bottom.toFloat()
        canvas.drawRect(mLineRect, paint)
        canvas.restore()
    }
}